improver/nbhood/use_nbhood.py

```diff
@@ -389,23 +404,21 @@ def process(self, cube):
         """
         # Mask out any NaNs in the neighbourhood data so that Iris ignores
         # them when calculating the weighted mean.
-        cube.data = ma.masked_invalid(cube.data)
+        cube.data = ma.masked_invalid(cube.data, copy=False)
         yname = cube.coord(axis="y").name()
         xname = cube.coord(axis="x").name()
 
-        if self.weights.shape != cube.shape:
-            # The input cube may have leading dimensions.
-            first_slice = next(
-                cube.slices([self.coord_masked, yname, xname], ordered=False)
-            )
-            self.renormalize_weights(first_slice)
-        else:
-            self.renormalize_weights(cube)
-        weights = self.weights.data
+        renormalize = True
+        if self.weights.shape == cube.shape:
+            weights = self.renormalize_weights(cube)
+            renormalize = False
 
         # Loop over any extra dimensions
         cubelist = iris.cube.CubeList([])
         for slice_3d in cube.slices([self.coord_masked, yname, xname]):
+            if renormalize:
+                weights = self.renormalize_weights(slice_3d)
+                renormalize = False
```

Comment:
I've changed the `renormalize_weights` method to return result rather than modify weights in place, since this operation depends on the input cube, but otherwise I kept the logic (nearly) the same. Still, I would expect this should be called on each slice rather than just the first one. Also, is condition checking `self.weights.shape` against `cube.shape` sufficient without comparing coordinates too? @fionaRust, what sort of weights shapes and renormalization is expected here?